cloud/pkg/leaderelection/leaderelection.go

```diff
@@ -0,0 +1,208 @@
+	readyzAdaptor.SetLeaderElection(leaderElector)
+	if err != nil {
+		klog.Errorf("couldn't create leader elector: %v", err)
+		return
+	}
+
+	// Start leaderElection until becoming leader, terminate program if leader lost or context.cancel
+	go leaderElector.Run(beehiveContext.GetContext())
+
+	// Monitor system signal and shutdown gracefully and it should be in main gorutine
+	core.GracefulShutdown()
+	return
+}
+
+// makeLeaderElectionConfig builds a leader election configuration. It will
+// create a new resource lock associated with the configuration.
+func makeLeaderElectionConfig(config componentbaseconfig.LeaderElectionConfiguration, client clientset.Interface, recorder record.EventRecorder) (*leaderelection.LeaderElectionConfig, error) {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get hostname: %v", err)
+	}
+	// add a uniquifier so that two processes on the same host don't accidentally both become active
+	id := hostname + "_" + string(uuid.NewUUID())
+
+	rl, err := resourcelock.New(config.ResourceLock,
+		config.ResourceNamespace,
+		config.ResourceName,
+		client.CoreV1(),
+		client.CoordinationV1(),
+		resourcelock.ResourceLockConfig{
+			Identity:      id,
+			EventRecorder: recorder,
+		})
+	if err != nil {
+		return nil, fmt.Errorf("couldn't create resource lock: %v", err)
+	}
+
+	return &leaderelection.LeaderElectionConfig{
+		Lock:          rl,
+		LeaseDuration: config.LeaseDuration.Duration,
+		RenewDeadline: config.RenewDeadline.Duration,
+		RetryPeriod:   config.RetryPeriod.Duration,
+		WatchDog:      nil,
+		Name:          "cloudcore",
+	}, nil
+}
+
+// Try to patch PodReadinessGate if program runs in pod
+func TryToPatchPodReadinessGate() error {
+	podname, isInPod := os.LookupEnv("CLOUDCORE_POD_NAME")
+	if isInPod == true {
+		namespace := os.Getenv("CLOUDCORE_POD_NAMESPACE")
+		klog.Infof("CloudCore is running in pod %v/%v, try to patch PodReadinessGate", namespace, podname)
+		//TODO: use specific clients
+		cli, err := utils.KubeClient()
+		if err != nil {
+			return fmt.Errorf("create kube client for patching podReadinessGate failed with error: %v", err)
+		}
+
+		//Creat patchBytes
+		getPod, err := cli.CoreV1().Pods(namespace).Get(podname, metav1.GetOptions{})
+		originalJSON, err := json.Marshal(getPod)
+		if err != nil {
+			return fmt.Errorf("failed to marshal modified pod %q into JSON: %v", podname, err)
+		}
+		//Todo: Read PodReadinessGate from CloudCore configuration or env
+		condition := corev1.PodCondition{Type: "kubeedge.io/CloudCoreIsLeader", Status: corev1.ConditionTrue}
+		podutil.UpdatePodCondition(&getPod.Status, &condition)
+		newJSON, err := json.Marshal(getPod)
+		patchBytes, err := strategicpatch.CreateTwoWayMergePatch(originalJSON, newJSON, corev1.Pod{})
+		if err != nil {
+			return fmt.Errorf("failed to create two way merge patch: %v", err)
+		}
+
+		//Try to patch
+		var isPatchSuccess = false
+		for i := 1; i <= 5; i++ {
```

Comment:
To avoid magic number, you can define a const e.g. `maxPodPatchRetry` in the beginning of the file.